<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Комментарий ревьюера</b>


Привет, Имран! Давай знакомиться! Меня зовут Дмитрий, и я буду проверять твой проект. Сразу предлагаю общение на «ты» 🙂, но если тебе это не комфортно, то дай знать, и мы перейдем на «вы». 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе погрузиться в увлекательный мир работы с данными и вырасти в крепкого профи. Это отдаленная цель. А ближайшая - сделать твою работу еще лучше )).
   
    
Все ключевые этапы в работе выполнены, и я вижу что с проектом ты справшяешься. Есть моменты, которые нужно доработать, но я уверен, у тебя все получится.
        
Расскажу как обычно проходит проверка:    
Бывают моменты, которые требуют пристального внимания. Комментарии по ним выделены <span style='background-color:#F7B3A4'> красным цветом </span> и обозначены значком 🛑. После их доработки проект будет принят.  🙂
 
<span style='background-color:#B7EBA7'> Зеленым цветом </span> и значком ✅ отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. Или советы «со звездочкой», которые помогут тебе в будущем.

<span style='background-color:#F9EDA6'>Жёлтым цветом </span> и значком ⚠️ выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).

Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента:</b></font>
   
Дмитрий, привет! Приятно познакомиться, конечно, можем на "ты". Огромное спасибо за советы и подсказки, а также за рекомендации!

</div>

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Комментарий ревьюера v2</b>

Привет! 🤝

Давай посмотрим на обновленый проект.
</div>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><span><a href="#Загрузим-и-ознакомимся-с-данными" data-toc-modified-id="Загрузим-и-ознакомимся-с-данными-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Загрузим и ознакомимся с данными</a></span></li><li><span><a href="#Исследуем-целевой-признак" data-toc-modified-id="Исследуем-целевой-признак-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Исследуем целевой признак</a></span></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#Загрузка-предобученной-модели-и-токенизация" data-toc-modified-id="Загрузка-предобученной-модели-и-токенизация-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Загрузка предобученной модели и токенизация</a></span></li><li><span><a href="#Создание-эмбеддингов" data-toc-modified-id="Создание-эмбеддингов-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Создание эмбеддингов</a></span></li><li><span><a href="#Обучим-и-протестируем-несколько-моделей" data-toc-modified-id="Обучим-и-протестируем-несколько-моделей-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обучим и протестируем несколько моделей</a></span></li><li><span><a href="#Попробую-немного-иной-подход-в-подборе-моделей-(старый-код-выше-в-маркдауне):" data-toc-modified-id="Попробую-немного-иной-подход-в-подборе-моделей-(старый-код-выше-в-маркдауне):-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Попробую немного иной подход в подборе моделей (старый код выше в маркдауне):</a></span></li><li><span><a href="#Проверяем-лучшую-модель-на-тестовой-выборке" data-toc-modified-id="Проверяем-лучшую-модель-на-тестовой-выборке-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Проверяем лучшую модель на тестовой выборке</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

**Привет!
Никак не получается достичь целевого показателя, не могу понять, в чем причина, многое перебробовал. Подскажи, пожалуйста, быть может есть неочевидная для меня ошибка.\
Балансировка классов только ухудшает показатели, а при переборе ГП все равно не хватает 2-3% до целевого показателя**

# Проект для «Викишоп» с использованием BERT

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Подготовка

Импортируем библиотеки:

In [39]:
import numpy as np
import pandas as pd
import torch
import transformers
from transformers import BertTokenizer, BertModel
from tqdm import notebook
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import (train_test_split, 
                                     cross_val_score,
                                       GridSearchCV,
                                       cross_validate)
from sklearn.metrics import f1_score, make_scorer
from sklearn.linear_model import LogisticRegression
from sklearn.utils import shuffle
from catboost import CatBoostClassifier
from numpy.random import RandomState
from lightgbm import LGBMClassifier
import re
from imblearn.over_sampling import SMOTE
from sklearn.dummy import DummyClassifier




<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично, все нужные библиотеки импортированы в начале ноутбука.Это хорошая практика.</div>

### Загрузим и ознакомимся с данными

In [2]:
data = pd.read_csv('toxic_comments.csv')

In [3]:
display(data.info(), data)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159292 entries, 0 to 159291
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   Unnamed: 0  159292 non-null  int64 
 1   text        159292 non-null  object
 2   toxic       159292 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 3.6+ MB


None

Unnamed: 0                                               text  toxic
0                0  Explanation\nWhy the edits made under my usern...      0
1                1  D'aww! He matches this background colour I'm s...      0
2                2  Hey man, I'm really not trying to edit war. It...      0
3                3  "\nMore\nI can't make any real suggestions on ...      0
4                4  You, sir, are my hero. Any chance you remember...      0
...            ...                                                ...    ...
159287      159446  ":::::And for the second time of asking, when ...      0
159288      159447  You should be ashamed of yourself \n\nThat is ...      0
159289      159448  Spitzer \n\nUmm, theres no actual article for ...      0
159290      159449  And it looks like it was actually you who put ...      0
159291      159450  "\nAnd ... I really don't think you understand...      0

[159292 rows x 3 columns]

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Данные загружены и есть первое знакомство.</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
А вот оборачивать результат работы метода  .info() в функцию print() или display() это лишнее. Метод  info() и сам напечатает тебе нужную информацию. А вот через print()/display() ты выводишь только то,что данный метод возвращает - т.е. None
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента:</b></font>
   
Да, согласен, насколько я знаю, display или print нужен в случаях, когда необходимо вывести несколько объектов/параметров на экран
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
 <b>v2</b>  

Обычно - да, но в случае info(), там print уже работает "под капотом". А вызывая принудительно display ты выводишь только строчку None    
    
</div>

Итак, в нашем датафрейме 3 столбца: Unnamed, text и toxic. Что необходимо сделать дальше:
1. Unnamed - похож на старые индексы, которые, возможно, остались после некоторых манипуляций с датафреймом. Нам он не понадобится, поэтому удалим его.
2. text - содержит непостредственно тексты комментариев, пропусков нет. Есть лишние символы /n, попробуем их заменить на пробелы.
3. toxic - признак токсичного комментария, также без пропусков.

In [4]:
data.drop('Unnamed: 0', axis = 1, inplace = True)

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
    
Да, все верно, столбец `'Unnamed: 0'` можно безболезненно удалить. Это "старый" индекс, который похоже забыли при пересохранении датасета.</div>

In [5]:
data.duplicated().sum()

0

In [6]:
data['text'].duplicated().sum()

0

In [7]:
data['text'] = data['text'].str.replace('\n', ' ')

In [8]:
data.head()

text  toxic
0  Explanation Why the edits made under my userna...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0
3  " More I can't make any real suggestions on im...      0
4  You, sir, are my hero. Any chance you remember...      0

In [12]:
def clean_text(text):
    text = text.lower()
    text = re.sub(r'<[^>]+>', '', text)  # Удаление HTML-тегов
    text = re.sub(r'http\S+|www\S+|https\S+', '', text, flags=re.MULTILINE)  # Удаление ссылок
    text = re.sub(r'\d+', '', text)  # Удаление чисел
    text = re.sub(r'[^a-zA-Z0-9\s.,\-:;\']', '', text)  # Удаление всех символов, кроме букв, цифр и пробелов
    text = text.strip()  # Удаление пробелов в начале и конце строки
    return text

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Учитывая что ты делаеш ьпредобрабтку для BERT я бы оставил еще знаки препинания, т.к. эта модель обучалась на оригинальных "сырых" текстах, и предобработку в данном случае есть смысл производить только для того, чтобы подчистить наши тексты от явного мусора</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента:</b></font>
   
Чуть подправил функцию, чтобы основные знаки препинания были сохранены</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
    <b>v2</b> 👍 </div>

In [13]:
data['text_cleaned'] = data['text'].apply(clean_text)

In [14]:
data.head()

text  toxic  \
0  Explanation Why the edits made under my userna...      0   
1  D'aww! He matches this background colour I'm s...      0   
2  Hey man, I'm really not trying to edit war. It...      0   
3  " More I can't make any real suggestions on im...      0   
4  You, sir, are my hero. Any chance you remember...      0   

                                        text_cleaned  
0  explanation why the edits made under my userna...  
1  d'aww he matches this background colour i'm se...  
2  hey man, i'm really not trying to edit war. it...  
3  more i can't make any real suggestions on impr...  
4  you, sir, are my hero. any chance you remember...

От ненужного столбца избавились, а также провели проверку на дубликаты - не обнаружены. Кроме этого, избавились от лишних символов в колонке text_cleaned

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 👍 </div>

### Исследуем целевой признак

In [15]:
data.pivot_table(index = 'toxic', aggfunc = 'count')

text  text_cleaned
toxic                      
0      143106        143106
1       16186         16186

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Молодец, исследован баланс классов. Это важная информация для задачи классификации.</div>

Видно явный дисбаланс классов, в соотношении примерно 1 к 9

## Обучение

Подготовим выборки features и target

In [17]:
features = data['text_cleaned']
target = data['toxic']

### Загрузка предобученной модели и токенизация

**Важно: весь код работы с BERT переведен в маркдаун во избежание повторных расчетов при перезапусках. Эмбеддинги были сохранены в файл и будут подгружаться вместо расчетов** 

In [18]:
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')

In [19]:
tokenizer = BertTokenizer.from_pretrained('unitary/toxic-bert')
model = BertModel.from_pretrained('unitary/toxic-bert').to(device)

tokenizer_config.json:   0%|          | 0.00/174 [00:00<?, ?B/s]

c:\Users\isara\AppData\Local\Programs\Python\Python311\Lib\site-packages\huggingface_hub\file_download.py:157: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\isara\.cache\huggingface\hub\models--unitary--toxic-bert. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/811 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/438M [00:00<?, ?B/s]

In [20]:
tokenized = features.apply(
    lambda x: tokenizer.encode(x, add_special_tokens=True, truncation = True))

max_len = 0
for i in tokenized.values:
    if len(i) > max_len:
        max_len = len(i)

padded = np.array([i + [0]*(max_len - len(i)) for i in tokenized.values])

attention_mask = np.where(padded != 0, 1, 0)

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 👍 </div>

### Создание эмбеддингов

In [21]:
batch_size = 100
embeddings = [] 
for i in notebook.tqdm(range((padded.shape[0] // batch_size)+1)):
    batch = torch.LongTensor(padded[batch_size * i:batch_size * (i+1)]).to(device)
    attention_mask_batch = torch.LongTensor(attention_mask[batch_size * i:batch_size * (i+1)]).to(device)

    with torch.no_grad():
        batch_embeddings = model(batch, attention_mask = attention_mask_batch)
    
    embeddings.append(batch_embeddings[0][:,0,:].cpu().numpy())

  0%|          | 0/1593 [00:00<?, ?it/s]

c:\Users\isara\AppData\Local\Programs\Python\Python311\Lib\site-packages\transformers\models\bert\modeling_bert.py:439: UserWarning: 1Torch was not compiled with flash attention. (Triggered internally at ..\aten\src\ATen\native\transformers\cuda\sdp_utils.cpp:455.)
  attn_output = torch.nn.functional.scaled_dot_product_attention(


Эмбеддинги созданы, объединим их в 1 массив, а также на всякий случай сохраним в файл.

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Чтобы код стабильно работал и на компьюетрах с GPU и без лучше использовать .to(device), а не .to('cuda') 

</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента:</b></font>
   
Поправил. Перенос расчетов на GPU был для меня настоящей победой с долгой битвой, поэтому просочились всякие странности)</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 <b>v2</b> да, это действительно не просто, зато как  прокачивает "инженерную интуицию" ))).

    
</div>

In [22]:
features_emb = np.concatenate(embeddings)
np.savetxt("features_emb_bert_toxic.csv", features_emb, delimiter = ',')

Итого, для создания батчей потребовалось почти 4 часа, сохраним результат в файл, чтобы при перезапуске кода не создавать батчи по новой

**UPD - 4 часа уходило на расчет только части датасета, с подключением GPU на полные модели BERT стало уходить ~45 минут на весь датасет. Но в файлик все равно сохраняю результат - импорт все же занимает менее минуты**

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 👍 </div>

### Обучим и протестируем несколько моделей

Используем признаки от bert, импорт признаков актуален при перезапуске проекта

features_emb = pd.read_csv('features_emb_clear_bert.csv', header=None)
features_emb.head()

Для начала подготовим обучающую и тестовую выборки

In [24]:
features_train, features_test, target_train, target_test = train_test_split(
    features_emb, target.values, test_size = .20, random_state = 12345)
display(features_train.shape, features_test.shape, target_train.shape, target_test.shape)


(127433, 768)

(31859, 768)

(127433,)

(31859,)

**Важно - прошлый код я перевел в маркдаун, т.к. решил структурировать побор моделей немного иначе. Новый расчет после старого кода.**

**Логистическая регрессия**

f1_scorer = make_scorer(f1_score)
state = RandomState(12345)

model_logreg = LogisticRegression()
log_param_grid = {
    'C': [0.1, 1, 5, 10],
    'max_iter': [50, 100, 300, 500, 1000],
    'solver': ['liblinear'],
    'random_state': [state]}

model_log_grid = GridSearchCV(model_logreg, log_param_grid, scoring = 'f1_macro')
model_log_grid.fit(features_train, target_train)

print(f'Best Parameters for Logistic Regression: {model_log_grid.best_params_}')
print(f'Best CV F1 Score for Logistic Regression: {model_log_grid.best_score_}')

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Метрика 'f1_macro' - это метрика для многоклассовой классификации, а мы  решаем задачу обычной бинарной классификации, поэтому пожалуйста используй обычную метрику f1</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger"> 
Не стоит раньше времени подглядывать в тестовую выборку. Вначале обучим все модели и выберем лучшую по оценкам на крос-валидации и только потом протестируем лучшую модель.</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger"> 
 Я не буду дублировать предыдущие два красных комментария, но пожалуйста исправь их везде ниже..</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента:</b></font>
   
Поправил везде. Выбор лучшей модели, а также ее тестирование в конце проекта
</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
    <b>v2</b> 👍 </div>

**Catboost:**

model_catboost = CatBoostClassifier()
catboost_param_grid = {
    'iterations': [500, 1000],
    'learning_rate': [0.05, 0.1, 0.3, 0.5],
    'verbose': [False]
}

model_cb_grid = GridSearchCV(model_catboost, catboost_param_grid, scoring = 'f1_macro')
model_cb_grid.fit(features_train, target_train)

print(f'Best Parameters for CatBoost: {model_cb_grid.best_params_}')
print(f'Best CV F1 Score for CatBoost: {model_cb_grid.best_score_}')

**LightGBM:**

model_lgbm = LGBMClassifier()
lgbm_param_grid = {
    'num_leaves': [50, 70, 100], 
    'learning_rate': [0.01, 0.1, 0.3, 0.5], 
    'random_state': [state],
    'verbose': [-1]
    }

model_lgbm_grid = GridSearchCV(model_lgbm, lgbm_param_grid, scoring = 'f1_macro')
model_lgbm_grid.fit(features_train, target_train)

print(f'Best Parameters for LGBM: {model_lgbm_grid.best_params_}')
print(f'Best CV F1 Score for LGBM: {model_lgbm_grid.best_score_}')


<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger"> 
И вот только сейчас, когда ты исследовал все модели нужно выбрать одну, самую лучшую, используя оценки полученные с помощью кросс-валидации и эту модель протестировать..</div>

### Попробую немного иной подход в подборе моделей (старый код выше в маркдауне):

In [26]:

f1_scorer = make_scorer(f1_score)
state = RandomState(12345)


models = [
    {
        'model': LogisticRegression(),
        'params': {
            'C': [0.1, 1, 5, 10],
            'max_iter': [300, 500, 1000],
            'solver': ['liblinear'],
            'random_state': [state]
        }
    },
    {
        'model': CatBoostClassifier(),
        'params': {
            'iterations': [500, 1000],
            'learning_rate': [0.05, 0.1, 0.3, 0.5],
            'verbose': [False]
        }
    },
    {
        'model': LGBMClassifier(),
        'params': {
            'num_leaves': [50, 70, 100], 
            'learning_rate': [0.01, 0.1, 0.3, 0.5], 
            'random_state': [state],
            'verbose': [-1]
        }
    }
    

]

In [33]:
for m in models:
    grid_search = GridSearchCV(estimator = m['model'], param_grid = m['params'], scoring = f1_scorer)
    grid_search.fit(features_train, target_train)
    best_model = grid_search.best_estimator_

    scores = cross_validate(best_model, features_train, target_train, scoring=f1_scorer, cv=5, return_train_score=True)
    
    print(f"Модель: {m['model'].__class__.__name__}")
    print(f"Лучшие параметры: {grid_search.best_params_}")
    print(f"Средний F1 на обучающей выборке: {np.mean(scores['train_score'])}")
    print(f"Средний F1 на валидационной выборке: {np.mean(scores['test_score'])}")
    print("-" * 30)

Модель: LogisticRegression
Лучшие параметры: {'C': 0.1, 'max_iter': 300, 'random_state': RandomState(MT19937) at 0x19C71867E40, 'solver': 'liblinear'}
Средний F1 на обучающей выборке: 0.9366401670139911
Средний F1 на валидационной выборке: 0.930258820705814
------------------------------
Модель: CatBoostClassifier
Лучшие параметры: {'iterations': 1000, 'learning_rate': 0.05, 'verbose': False}
Средний F1 на обучающей выборке: 0.9868495068477792
Средний F1 на валидационной выборке: 0.9298056465306754
------------------------------
Модель: LGBMClassifier
Лучшие параметры: {'learning_rate': 0.1, 'num_leaves': 70, 'random_state': RandomState(MT19937) at 0x19C71867E40, 'verbose': -1}
Средний F1 на обучающей выборке: 0.9995271550816754
Средний F1 на валидационной выборке: 0.9290320615426768
------------------------------


In [34]:
best_model

LGBMClassifier(num_leaves=70,
               random_state=RandomState(MT19937) at 0x19B37A27340, verbose=-1)

**Сохраним результаты подбора моделей:**

Модель: LogisticRegression\
Лучшие параметры: {'C': 0.1, 'max_iter': 300, 'random_state': RandomState(MT19937) at 0x19C71867E40, 'solver': 'liblinear'}\
Средний F1 на обучающей выборке: 0.9366401670139911\
Средний F1 на валидационной выборке: 0.930258820705814\

------------------------------

Модель: CatBoostClassifier\
Лучшие параметры: {'iterations': 1000, 'learning_rate': 0.05, 'verbose': False}\
Средний F1 на обучающей выборке: 0.9868495068477792\
Средний F1 на валидационной выборке: 0.9298056465306754\

------------------------------

Модель: LGBMClassifier\
Лучшие параметры: {'learning_rate': 0.1, 'num_leaves': 70, 'random_state': RandomState(MT19937) at 0x19C71867E40, 'verbose': -1}\
Средний F1 на обучающей выборке: 0.9995271550816754\
Средний F1 на валидационной выборке: 0.9290320615426768\


**Результаты моделей оказались очень близки друг к другу, но лучшей все же оказалась LGBMClassifier. В целом, имеет смысл проверить все модели на тестовой выборке**


### Проверяем лучшую модель на тестовой выборке

In [36]:
# Проверяем лучшую модель на тестовой выборке:
print('F1-мера для LGBM Тест:', f1_score(target_test, best_model.predict(features_test)))

F1-мера для LGBM Тест: 0.9356852950359038


Получилось - результат намного выше целевого значения, F1-мера для лучшей модели составила: 0.9356852950359038

Ради интереса проверим и оставшиеся модели:

In [38]:
logreg_best = LogisticRegression(C = 0.1, max_iter = 300, random_state = state, solver = 'liblinear')
logreg_best.fit(features_train, target_train)

catboost_best = CatBoostClassifier(iterations = 1000, learning_rate = 0.05, verbose = False)
catboost_best.fit(features_train, target_train)

print('F1-мера для Логистической регрессии Тест:', f1_score(target_test, logreg_best.predict(features_test)))
print('F1-мера для CatBoost Тест:', f1_score(target_test, catboost_best.predict(features_test)))

F1-мера для Логистической регрессии Тест: 0.940405616224649
F1-мера для CatBoost Тест: 0.9352473084724606


Итого, на тестовой выборке лучшей оказалось модель Логистическая регрессия, но необходимо отметить - результаты всех моделей оказались очень близки.

Теперь сравним результаты наших моделей с dummy:

In [41]:
dummy = DummyClassifier(strategy = "stratified")
dummy.fit(features_train, target_train)
print('F1-мера для Dummy Тест:', f1_score(target_test, dummy.predict(features_test)))


F1-мера для Dummy Тест: 0.0972200652275198


В целом, результат ожидаемый.

## Выводы


В рамках задачи по проекту можно отметить следующее:

1. В работу был принят датасет из 159292 текстов, по которым были проставлены признаки токсичности. В рамках предобработки, был удален лишний столбец "unnamed: 0", а также удалены лишние символы из текстов. Пропуски в данных и дубликаты не были обнаружены. Также был выявлен сильный дисбаланс классов - соотношение составило примерно 1 к 9, в польу 0-класса. Забегая вперед, можно отметить, что балансировка классов в данном случае не пригодилась.

2. С помощью предобученной модели BERT - unitary/toxic-bert была произведена токенизация текстов и создание эмбеддингов. Здесь можно отметить, что base-bert, distil-, tiny- bert не приводили к нужным результатам - здесь требовалась конфигурация, направленная на токсичные тексты. В итоге удалось создать нужные эмбеддинги для определния токсичных комментариев.

3. С помощью GridSearchCV и cross_validate были обучены модели с подборобом гиперпараметров: LogisticRegression, CatBoostClassifier, LGBMClassifier. Лучшей моделью автоматически был определен LGBMClassifier. Но, результаты всех моделей были близки.

4. На тестовой выборке была протестирована лучшая модель, ее результат составил f1 = 0.9356852950359038. При этом,  f1 Логистической регрессии = 0.940405616224649, f1 CatBoost = 0.9352473084724606, иными словами, результаты на тестовой выборке оказались также очень близки.

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Модели обучены
- [x]  Значение метрики *F1* не меньше 0.75
- [x]  Выводы написаны

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> комментарий ревьюера</b>
Ты правильно выполнил самую сложную часть и получил признаки из  текста с использованием предобученной модели BERT.
    
Метрика получилась достаточно низкая. Что же делать и как её увеличить? К сожалению врядли получится достичь нужного значения метрики, даже увеличив объем данных (хотя шансы приблизиться к отметке 0,75 есть, если использовать все данные и долго подбирать гиперпараметры моделей). Но почему? Это же хваленый трансформер, про который сейчас говорят из каждого утюга.  Дело в том, что ты используешь "обычный" BERT, обученный на обычных текстах и похоже признаки, которые мы получаем с его помощью плохо подходят для решения нашей задачи. Можно дообучить эту модель на наших данных, но это сложно и выходит за рамки проекта.  Самый доступный выход - найти модель, уже обученную на определение токсичности. И такие модели легко находятся на HugingFace по ключевому слову toxic.  Мой лидер - unitary/toxic-bert. С этой моделью легко получается 0,9 или даже 0,95.
     
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента:</b></font>
Да уж, я полагал, что полновесный BERT умеет в т.ч. и в токсичность на должном уровне, даже идеи не было поискать toxic - friendly варианты BERT. Показательный случай на будущее, большое спасибо, что направил!

UPD - результат unitary/toxic-bert впечатляет, стремится к 0.95 по всем протестированным моделям!
</div>

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Заключительный комментарий v2</b>

Ну уж если быть совсем "честным", то было бы неплохо проверить (в карточке модели), что авторы этой модели не показывали ей наш датасет... Такая проблема с утечками последнее время часто возникает с большими моделями, в обучении которых попадают данные из разных бенчмарков.
    
 ---------------   
    
Проект принят. Поздравляю и желаю дальнейших успехов!
</div>

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> рекомендации по доп. материалам</b>
Если решишь погрузиться в область работы с текстами, очень советую несколько продвинутых бесплатных курсов.
    
   - Отличный бесплатный курс от Школы глубокого обучения МФТИ (https://stepik.org/org/dlschool), старт курса каждые пол года. Два семестра, один по основам и компьютерному зрению, второй по обработке естественного языка. Проходить нужно именно в таком порядке,т.к. почти весь современный NLP построен на нейронках.
    
   - "Нейронные сети и компьютерное зрение" от Samsung Research Russia (https://stepik.org/course/50352/syllabus). Есть также продолжение по NLP.  
   - Трек NLP от сообщества ODS https://ods.ai/tracks/nlp-course-autumn-22
    
    
А если на тебя произвели впечатление возможности ChatGPT и хочешь попробовать использовать возможности больших языковых моделей для решения своих задач, могу порекомендовать следующие курсы (первые две ссылки на английском, зато последняя подборка лекций на русском).
    
  - https://www.coursera.org/learn/generative-ai-with-llms  (можно прослужать бесплатно)
  - Короткие курсы на сайте https://www.deeplearning.ai/short-courses/  Самый свежак - как писать промпты, как использовать LLM для создания собственных приложений (например как создать бота на основе ChatGPT, который будет отвечать на вопросы по внутренней документации компании).  
  - https://www.promptingguide.ai/introduction/settings
  - https://www.youtube.com/watch?v=l-l82uNwyu8&list=PLy6K3_Hx-udj6n1S88Vslyw2QVxSXLP2c    
    
    
    
    
Также можно обратиться к туториалу по использованию языковых моделей на HugingFace. Обрати внимание, для больше чем половины материалов доступен перевод на русский, хотя можно читать и на языке оригинала. 
    
  https://huggingface.co/learn/nlp-course   
 
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента:</b></font>
Однозначно да, на курсе Практикума материалов весьма мало по теме, поэтому обязательно нужно дополнительное изучение, спасибо за подборку!
</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
    <b>v2</b> 🙏 </div>